cabal-install/Distribution/Client/Outdated.hs

```diff
@@ -0,0 +1,134 @@
+outdated :: Verbosity -> OutdatedFlags -> RepoContext
+         -> Compiler -> Platform
+         -> IO ()
+outdated verbosity outdatedFlags repoContext comp platform = do
+  sourcePkgDb <- IndexUtils.getSourcePackages verbosity repoContext
+  let freeze   = fromFlagOrDefault False (outdatedFreeze outdatedFlags)
+      exitCode = fromFlagOrDefault False (outdatedExitCode outdatedFlags)
+      ignore   = S.fromList (outdatedIgnore outdatedFlags)
+      minor    = S.fromList (outdatedMinor outdatedFlags)
+      pkgIndex = packageIndex sourcePkgDb
+  deps <- if freeze
+          then depsFromFreezeFile verbosity
+          else depsFromPkgDesc    verbosity comp platform
+  let outdatedDeps = listOutdated deps pkgIndex ignore minor
+  notice verbosity ("Outdated dependencies: "
+                     ++ intercalate ", "
+                        (map (\(d, v) -> display d
+                               ++ " (latest: " ++ display v ++ ")") outdatedDeps))
+  if (exitCode && (not . null $ outdatedDeps))
+    then exitFailure
+    else return ()
+
+-- | Read the list of dependencies from the freeze file.
+depsFromFreezeFile :: Verbosity -> IO [Dependency]
+depsFromFreezeFile verbosity = do
+  cwd        <- getCurrentDirectory
+  userConfig <- loadUserConfig verbosity cwd Nothing
+  let ucnstrs = map fst . configExConstraints . savedConfigureExFlags $ userConfig
+      deps    = mapMaybe (packageConstraintToDependency . userToPackageConstraint)
+                ucnstrs
+  debug verbosity "Reading the list of dependencies from the freeze file"
+  return deps
+
+-- | Read the list of dependencies from the package description.
+depsFromPkgDesc :: Verbosity -> Compiler  -> Platform -> IO [Dependency]
+depsFromPkgDesc verbosity comp platform = do
+  cwd  <- getCurrentDirectory
+  path <- tryFindPackageDesc cwd
+  gpd  <- readPackageDescription verbosity path
+  let cinfo = compilerInfo comp
+      epd = finalizePD [] defaultComponentRequestedSpec
+            (const True) platform cinfo [] gpd
+  case epd of
+    Left _        -> die "finalizePD failed"
+    Right (pd, _) -> do
+      let bd = buildDepends pd
+      debug verbosity
+        "Reading the list of dependencies from the package description"
+      return bd
+
+-- | Find all outdated dependencies.
+listOutdated :: [Dependency] -> PackageIndex UnresolvedSourcePackage
+              -> S.Set PackageName -> S.Set PackageName
+              -> [(Dependency, Version)]
+listOutdated deps pkgIndex ignore minor =
+  mapMaybe isOutdated $ map simplifyDependency deps
+  where
+    isOutdated :: Dependency -> Maybe (Dependency, Version)
+    isOutdated dep
+      | depPkgName dep `S.member` ignore = Nothing
+      | otherwise                      =
+          let this   = map packageVersion $ lookupDependency pkgIndex dep
+              latest = lookupLatest dep
+          in (\v -> (dep, v)) `fmap` isOutdated' this latest
+
+    isOutdated' :: [Version] -> [Version] -> Maybe Version
```

Comment:
OK, I hope sortedness of lookupDependency output is actually an invariant we can rely on.